src/librustc_typeck/check/mod.rs

```diff
@@ -3408,6 +3408,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
             return self.tcx.types.err;
         };
 
+        if self.tcx.sess.features.borrow().non_exhaustive {
+            let is_local = variant.did.is_local();
+
+            if !is_local {
+                let descr = self.tcx.adt_def(variant.did).variant_descr();
+
+                // Prohibit struct expressions when non exhaustive flag is set.
+                span_err!(self.tcx.sess, expr.span, E0639,
+                          "cannot create non-exhaustive {}",
```

Comment:
Nit: `"cannot create non-exhaustive {} using struct expression"`?
I mean you can obtain non-exhaustive structs using other means, e.g. function calls.

Reply:
Fixed.